kore/src/Kore/Step/Rule/Combine.hs

```diff
@@ -0,0 +1,104 @@
+{-|
+Copyright   : (c) Runtime Verification, 2019
+License     : NCSA
+-}
+
+module Kore.Step.Rule.Combine
+    ( mergeRulesPredicate
+    ) where
+
+import Data.Set (Set)
+import qualified Data.Set as Set
+import qualified Data.Map as Map
+import qualified Data.List as List
+
+import Kore.Predicate.Predicate
+    ( makeCeilPredicate
+    , makeMultipleAndPredicate
+    )
+import qualified Kore.Predicate.Predicate as Syntax
+    ( Predicate
+    )
+import Kore.Step.Rule
+    ( RewriteRule (RewriteRule), RulePattern (RulePattern), refreshRulePattern
+    )
+import qualified Kore.Step.Rule as Rule (freeVariables)
+import qualified Kore.Step.Rule as Rule.DoNotUse
+import Kore.Internal.TermLike (mkAnd)
+import Kore.Internal.Variable (InternalVariable)
+import Kore.Attribute.Pattern.FreeVariables (FreeVariables(FreeVariables))
+import Kore.Variables.UnifiedVariable (UnifiedVariable)
+import Kore.Substitute (SubstitutionVariable)
+
+{-
+Given a list of rules
+
+@
+L1 -> R1
+L2 -> R2
+...
+Ln -> Rn
+@
+
+returns a predicate P such that applying the above rules in succession
+is the same as applying @(L1 and P) => Rn@.
+
+See docs/2019-09-09-Combining-Rewrite-Axioms.md for details.
+-}
+mergeRulesPredicate
+    :: SubstitutionVariable variable
+    => [RewriteRule variable]
+    -> Syntax.Predicate variable
+mergeRulesPredicate rules =
+    makeMultipleAndPredicate
+    $ map mergeRulePair
+    $ makeConsecutivePairs
+    $ renameRulesVariables rules
+
+makeConsecutivePairs :: [a] -> [(a, a)]
+makeConsecutivePairs [] = []
+makeConsecutivePairs [_] = []
+makeConsecutivePairs (a1 : a2 : as) = (a1, a2) : makeConsecutivePairs (a2 : as)
+
+mergeRulePair
+    :: InternalVariable variable
+    => (RewriteRule variable, RewriteRule variable)
+    -> Syntax.Predicate variable
+mergeRulePair
+    ( RewriteRule RulePattern {right = right1, ensures = ensures1}
+    , RewriteRule RulePattern {left = left2, requires = requires2}
+    )
+  =
+    makeMultipleAndPredicate
+        [ makeCeilPredicate (mkAnd right1 left2)
+        , ensures1
+        , requires2
+        ]
```

Comment:
Given:

```
l1 => r1 requires r1 ensures e1
l2 => r2 requires r2 ensures e2
```

This seems to build:

```
AND [ \ceil(\and(r1, l2)) , e1 , r2 ]
```

I'm surprised that `l1` and `r2` don't show up.

I guess I would expect something more like:

```
l1 => r2 requires AND [ \ceil(\and(r1, l2)) , e1 , r2 ]
```

Reply:
l1 and rn will be added last; the intermediary ones would become r1s and l2s.

Reply:
This only builds the predicate, I thought it is better to wait until we actually simplify the predicate to build the rule.

Reply:
Ah ok. In that case I guess I would suggest a different name than `mergeRulePair`, but I'm not very familiar with your style guides/naming schemes, so up to you.

Thanks for clarifying!